pmlc/conversion/pxa_to_affine/pxa_to_affine.cc

```diff
@@ -57,11 +57,47 @@ struct LoweringBase : public OpConversionPattern<OpType> {
   LogicalResult match(Operation *op) const override { return mlir::success(); }
 };
 
-struct AffineParallelOpConversion : public LoweringBase<AffineParallelOp> {
-  explicit AffineParallelOpConversion(MLIRContext *ctx) : LoweringBase(ctx) {}
+// This pattern removes affine.parallel ops with no induction variables
+struct AffineParallelRank0Remover
+    : public mlir::OpRewritePattern<AffineParallelOp> {
+  using mlir::OpRewritePattern<AffineParallelOp>::OpRewritePattern;
 
-  void rewrite(AffineParallelOp op, ArrayRef<Value> operands,
-               ConversionPatternRewriter &rewriter) const override {
+  explicit AffineParallelRank0Remover(MLIRContext *ctx)
+      : OpRewritePattern(ctx) {}
+
+  LogicalResult
+  matchAndRewrite(AffineParallelOp op,
+                  mlir::PatternRewriter &rewriter) const override {
+    // Check that there are no induction variables
+    if (op.lowerBoundsMap().getNumResults() != 0)
+      return mlir::failure();
+    // Remove the affine.parallel wrapper, retain the body in the same location
+    auto &parentOps = rewriter.getInsertionBlock()->getOperations();
+    auto &parallelBodyOps = op.region().front().getOperations();
+    parentOps.splice(mlir::Block::iterator(op), parallelBodyOps,
+                     parallelBodyOps.begin(), std::prev(parallelBodyOps.end()));
+    // Replace outputs with values from yield
+    auto termIt = std::prev(parallelBodyOps.end());
+    for (size_t i = 0; i < op.getNumResults(); i++) {
+      op.getResult(i).replaceAllUsesWith(termIt->getOperand(i));
+    }
+    rewriter.eraseOp(op);
+    return mlir::success();
+  }
+};
+
+struct AffineParallelOpConversion
+    : public OpConversionPattern<AffineParallelOp> {
```

Comment:
The other uses OpRewritePattern, maybe we should be consistent.

Reply:
The `AffineParallelRank0Remover` using `OpRewritePattern` is the inconsistent one; everything else uses `OpConversionPattern` via `LoweringBase`. I had originally thought of the rank 0 remover possibly seeing additional use outside of lowering, hence the use of a rewrite pattern, but we don't currently do that so perhaps I should investigate switching it to an OpConversionPattern?